waiter/src/waiter/auth/oauth.clj

```diff
@@ -0,0 +1,166 @@
+  (authenticate [this request]
+    "Handles the response from the OAuth provider.  Returns an async response."))
+
+(defn make-uri
+  "Takes a URI and adds a formatted query-string from query parameters in a map."
+  [uri query-params]
+  (str uri "?"
+       (->> query-params
+            (map (fn [[name value]]
+                   (str name "=" (UrlEncoded/encodeString value))))
+            (str/join "&"))))
+
+(defn make-location
+  "Given a path and query-string, return the value for the Location header
+  after successful authentication.
+  Assumes the URL fragment (hash) is passed at the end of query-string under a special
+  query parameter HASH-QUERY-PARAM-NAME."
+  [{:keys [path query-string]}]
+  (let [query-string' (when query-string
+                        (str \? query-string))
+        hash (when (and query-string' (str/includes? query-string' (str HASH-QUERY-PARAM-NAME \=)))
+               (-> (subs query-string' (+ (str/index-of query-string' (str HASH-QUERY-PARAM-NAME \=))
+                                          (count (str HASH-QUERY-PARAM-NAME \=))))
+                   (UrlEncoded/decodeString)))
+        query-string'' (if (and query-string' hash)
+                         (subs query-string' 0 (dec (str/index-of query-string' (str HASH-QUERY-PARAM-NAME \=))))
+                         query-string')]
+    (str \/ path query-string'' hash)))
+
+(defn provider-list-handler
+  "Responds with a login page listing available OAuth providers."
+  [providers {:keys [query-string request-method] {:keys [path]} :route-params}]
+  (case request-method
+    :get {:body (template/eval (slurp (io/resource "web/oauth.html"))
+                               {:providers providers
+                                :display-name (fn [provider] (display-name provider))
+                                :hash-query-param-name HASH-QUERY-PARAM-NAME
+                                :path path
+                                :query-string query-string})
+          :headers {"content-type" "text/html"}}
+    (throw (ex-info "Invalid request method"
+                    {:request-method request-method
+                     :status 405}))))
+
+(defn match-oauth-route
+  "Match an OAuth route based upon a uri."
+  [uri]
+  (let [route-map ["/waiter-auth/oauth" {["/providers/" [#".*" :path]] :provider-list
+                                         "/" {[:provider "/redirect/" [#".*" :path]] :redirect
+                                              [:provider "/authenticate"] :authenticate}}]]
+    (bidi/match-route route-map uri)))
+
+(defn wrap-oauth
+  "Middleware for handling OAuth."
+  [handler-fn run-as-user password providers]
+  (fn [{:keys [headers query-string uri] :as request}]
+    (try
+      (let [{:keys [handler route-params] :as match} (match-oauth-route uri)
+            request' (assoc request :route-params route-params)
+            provider-fn (fn [] (-> route-params :provider keyword providers))]
+        (case handler
```

Comment:
Can we please include counters on these handlers so that they show up in waiter requests, e.g. `(counters/inc! (metrics/waiter-counter "requests" "oauth" (name handler)))`?